libraries/AP_Beacon/AP_Beacon.cpp

```diff
@@ -225,6 +226,75 @@ uint32_t AP_Beacon::beacon_last_update_ms(uint8_t beacon_instance) const
     return beacon_state[beacon_instance].distance_update_ms;
 }
 
+// create fence boundary points
+bool AP_Beacon::update_boundary_points()
+{
+    // we need three points at least to create fence
+    // if already tried to create boundary, return false
+    if (!device_ready() || num_beacons < AP_BEACON_MINIMUM_FENCE_BEACONS || boundary_num_beacons == num_beacons) {
+        return false;
+    }
+
+    // create boundary (5 points with a closing point) at once
+    if (!boundary_create_attempted) {
+        uint32_t array_size = (AP_BEACON_MAX_BEACONS + 1) * sizeof(Vector2f);
+        if (hal.util->available_memory() >= 100U + array_size) {
+            _boundary = (Vector2f *)calloc(1, array_size);
+        }
+        boundary_create_attempted = true;
+    }
+
+    // check if boundary is available
+    if (_boundary == nullptr) {
+        return false;
+    } else {
+        boundary_num_beacons = num_beacons;
+    }
+
+    // accumulate beacon points
+    for (uint8_t index = 0; index < num_beacons; index++) {
+        // if a beacon is not healthy, delete _boundary to renew array
+#if CONFIG_HAL_BOARD != HAL_BOARD_SITL
+        if (!beacon_healthy(index)) {
+            delete _boundary;
+            _boundary = nullptr;
+            return false;
+        }
+#endif
+        Vector3f point_3d = beacon_position(index);
+        Vector2f point_2d(point_3d.x, point_3d.y);
+
+        _boundary[index] = point_2d;
+    }
+
+#if CONFIG_HAL_BOARD != HAL_BOARD_SITL
+    // reorder points to create fence region by Pozyx anchor
+    // swap 3rd point and 4th point if 4 points exist
+    // http://ardupilot.org/copter/docs/common-pozyx.html
+    if (_type == AP_BeaconType_Pozyx) {
```

Comment:
I think my previous comment wasn't clear enough. What I meant is that I think the frontend shouldn't be doing this for itself as it means it needs to be aware of details of each supported backend, so either:

- each backend creates the boundary array (basically the `update_boundary_points` method would only be defined in each backend and not in the frontend class)
- each backend is given the boundary array and it can change it as needed

I prefer the first as it looks a better architecture to me. The current method, for example, assumes it can create a boundary for the Marvelmind backend, but I think that it's not guaranteed it can form a boundary just based on the beacon indexes. Another example is that this checks for number of beacons to be different, but you can have the same number of beacons and the boundary needing to recalculated (for example, if a beacon is unhealthy).